### PR TITLE
add new fingerprints to dupe detection service

### PR DESCRIPTION
**Pull request by georgkcode0**
task description: https://freedcamp.com/view/2956207/tasks/panel/task/41229519

### CODE REVIEW

File: dupedetection/service.go [hunk 1]
```diff
@@ -0,0 +1,279 @@
...
+	"github.com/pastelnetwork/gonode/metadb/rqlite/db"
+	"github.com/pastelnetwork/gonode/p2p"
+	"github.com/pastelnetwork/gonode/pastel"
+	"github.com/sbinet/npyio"
+)
+
+const (
+	synchronizationIntervalSec    = 2
+	synchronizationTimeoutSec     = 60
+	runTaskIntervalMin            = 2
```
Comment (pjenys0) [inline dupedetection/service.go:24]: don't really think 2 minutes are reasonable.
maybe a half of hour?
could you double check with Jeff?

File: dupedetection/service.go [hunk 2]
```diff
@@ -0,0 +1,279 @@
...
+	}
+
+	for i := 0; i < len(artRegTickets); i++ {
+		if artRegTickets[i].Height <= lastestFp.NumberOfBlock {
+			continue
+		}
+
+		artTicketData, err := pastel.DecodeArtTicket(artRegTickets[i].RegTicketData.ArtTicket)
+		if err != nil {
+			return errors.Errorf("failed to decode reg ticket: err %w", err)
```
Comment (pjenys0) [inline dupedetection/service.go:210]: if error, should continue with next ticket

File: dupedetection/service.go [hunk 3]
```diff
@@ -0,0 +1,279 @@
...
+		}
+
+		fingerprintsHash := artTicketData.AppTicketData.FingerprintsHash
+		fingerprintByte, err := s.p2pClient.Retrieve(ctx, base58.Encode(fingerprintsHash))
+		if err != nil {
+			return errors.Errorf("failed to retrieve fingerprint, err: %w", err)
```
Comment (pjenys0) [inline dupedetection/service.go:216]: same if error, should continue with next ticket

File: dupedetection/service.go [hunk 4]
```diff
@@ -0,0 +1,279 @@
...
+		}
+
+		fingerprint, err := pastel.FingerprintFromBytes(fingerprintByte)
+		if err != nil {
+			return errors.Errorf("failed to convert fingerprint, err: %w", err)
```
Comment (pjenys0) [inline dupedetection/service.go:221]: same if error, should continue with next ticket?

File: dupedetection/service.go [hunk 5]
```diff
@@ -0,0 +1,279 @@
...
+		}
+
+		size := len(fingerprint)
+		if size != fingerprintSizeModel {
+			return errors.Errorf("invaild size fingerprint, size: %d", size)
```
Comment (pjenys0) [inline dupedetection/service.go:226]: same if error, should continue with next ticket

File: dupedetection/service.go [hunk 6]
```diff
@@ -0,0 +1,279 @@
...
+		if err := s.storeFingerprint(ctx, &dupeDetectionFingerprints{
+			Sha256HashOfArtImageFile:           hex.EncodeToString(fingerprintsHash),
+			Model1ImageFingerprintVector:       model1ImageFingerprintVector,
+			Model2ImageFingerprintVector:       model2ImageFingerprintVector,
+			Model3ImageFingerprintVector:       model3ImageFingerprintVector,
+			Model4ImageFingerprintVector:       model4ImageFingerprintVector,
+			NumberOfBlock:                      artRegTickets[i].Height,
+			DatetimeFingerprintAddedToDatabase: time.Now(),
+		}); err != nil {
+			return errors.Errorf("failed to store fingerprint, err: %w", err)
```
Comment (pjenys0) [inline dupedetection/service.go:255]: same if error, should continue with next ticket

File: dupedetection/service.go [hunk 7]
```diff
@@ -0,0 +1,279 @@
...
+
+	fp4, err := encodeFloat2Npy(input.Model4ImageFingerprintVector)
+	if err != nil {
+		return err
+	}
+
+	statement := fmt.Sprintf(insertFingerprintStatement, input.Sha256HashOfArtImageFile, fp1, fp2, fp3, fp4)
+	_, err = s.db.ExecuteStringStmt(statement)
+	if err != nil {
+		fmt.Printf("failed to insert fingerprint record: %s", err.Error())
```
Comment (pjenys0) [inline dupedetection/service.go:187]: should use log module than fmt.Printf

File: dupedetection/service.go [hunk 8]
```diff
@@ -0,0 +1,279 @@
...
+				return nil
+			}
+		case <-timeoutCh:
+			return errors.Errorf("timeout")
+		}
+	}
+}
+
+func (s *service) getLatestFingerprint(ctx context.Context) (*dupeDetectionFingerprints, error) {
+	row, err := s.db.QueryStringStmt(getLatestFingerprintStatement)
```
Comment (pjenys0) [inline dupedetection/service.go:110]: in case of database is too big, query all items seem not well?
anw, be fine now, but should improve it later

Comment (georgkcode0) [reply dupedetection/service.go:110]: noted

File: dupedetection/service.go (hunk 1 above)
Comment (georgkcode0) [reply dupedetection/service.go:24]: okay. let me check

**Review by danileng20 — APPROVED** [review]

**Review by pjenys0 — APPROVED** [review]

**Pull request merged**